### PR TITLE
Thread safe clean up of LocalNodeModeListeners

### PR DESCRIPTION
**Pull request by jaymode**
This commit continues on the work in #59801 and makes other
implementors of the `LocalNodeMasterListener` interface thread safe in
that they will no longer allow the callbacks to run on different
threads and possibly race each other. This also helps address other
issues where these events could be queued to wait for execution while
the service keeps moving forward thinking it is the master even when
that is not the case.

In order to accomplish this, the `LocalNodeMasterListener` no longer has
the `executorName()` method to prevent future uses that could encounter
this trappy behavior.

Each use was inspected and if the class was also a
`ClusterStateListener`, the implementation of `LocalNodeMasterListener`
was removed in favor of a single listener that combined the logic. A
single listener is used and there is currently no guarantee on execution
order between `ClusterStateListener`s and `LocalNodeMasterListener`s,
so a future change there could cause undesired consequences. For other
classes, the implementations of the callbacks were inspected and if the
operations were lightweight, the overriden `executorName` method was
removed to use the default, which runs on the same thread.

### CODE REVIEW

**Review by original-brownbear — APPROVED** [review]
LGTM, nice cleanup :)

**Review by dakrone — APPROVED** [review]
LGTM too, thanks Jay

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java [hunk 1]
```diff
@@ -111,9 +111,6 @@ public InternalClusterInfoService(Settings settings, ClusterService clusterServi
         clusterSettings.addSettingsUpdateConsumer(INTERNAL_CLUSTER_INFO_UPDATE_INTERVAL_SETTING, this::setUpdateFrequency);
         clusterSettings.addSettingsUpdateConsumer(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_THRESHOLD_ENABLED_SETTING,
                                                   this::setEnabled);
-
-        // listen for state changes (this node starts/stops being the elected master, or new nodes are added)
-        clusterService.addListener(this);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java:116]: Sorry for the conflicts with #59880, but I think we need this line still.

**Review by DaveCTurner — COMMENTED** [review]
Looks like a merge issue here, see comment below.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java:116]: Oh wait, no I see this got moved to `Node`. That's nicer indeed but I expect `InternalClusterInfoServiceSchedulingTests` need adjusting too.

File: server/src/main/java/org/elasticsearch/cluster/LocalNodeMasterListener.java [hunk 2]
```diff
@@ -37,16 +39,20 @@
...
-     * or {@link org.elasticsearch.threadpool.ThreadPool.Names#MANAGEMENT MANAGEMENT}).
+     * executor; for example management operations would use {@link org.elasticsearch.threadpool.ThreadPool.Names#MANAGEMENT MANAGEMENT}).
+     *
+     * <em>Note</em>: when using an executor with more than one thread the callbacks <i>could</i> run concurrently, so care
+     * should be taken to ensure proper locking is used for correctness.
      *
      * @return The name of the executor that will run the callbacks of this listener.
      */
-    String executorName();
+    default String executorName() {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/LocalNodeMasterListener.java:53]: This looks never to be overridden, can we inline it (and simplify `threadPool.executor(SAME).execute(r)` to `r.run()` in the `ClusterApplierService`?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java [hunk 3]
```diff
@@ -131,11 +130,6 @@ public void stop() {
         logger.debug("ML memory tracker stopped");
     }
 
-    @Override
-    public String executorName() {
```
Comment (DaveCTurner) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java:135]: We use this executor to run `MlMemoryTracker#refresh` and `MlMemoryTracker #iterateAnomalyDetectorJobTasks`, are you sure it's ok to run them on the `SAME` thread now?

**Review by DaveCTurner — COMMENTED** [review]
Couple more comments.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java (hunk 1 above)
Comment (jaymode) [reply server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java:116]: No need to apologize. Addressed in https://github.com/elastic/elasticsearch/pull/59932/commits/59a7feee7cc2ec9ce13f353a3af798e6eaa06116

File: server/src/main/java/org/elasticsearch/cluster/LocalNodeMasterListener.java (hunk 2 above)
Comment (jaymode) [reply server/src/main/java/org/elasticsearch/cluster/LocalNodeMasterListener.java]: Addressed in https://github.com/elastic/elasticsearch/pull/59932/commits/63be6ab56326f0a496a33588730130aa227bb45f

File: server/src/main/java/org/elasticsearch/cluster/LocalNodeMasterListener.java [hunk 4]
```diff
@@ -33,20 +33,5 @@
...
-     * the work done is the callbacks above is inexpensive, this value may be
-     * {@link org.elasticsearch.threadpool.ThreadPool.Names#SAME SAME} (indicating that the callbacks will run on the same thread
-     * as the cluster state events are fired with). On the other hand, if the logic in the callbacks are heavier and take
-     * longer to process (or perhaps involve blocking due to IO operations), prefer to execute them on a separate more appropriate
-     * executor (eg. {@link org.elasticsearch.threadpool.ThreadPool.Names#GENERIC GENERIC}
-     * or {@link org.elasticsearch.threadpool.ThreadPool.Names#MANAGEMENT MANAGEMENT}).
-     *
-     * @return The name of the executor that will run the callbacks of this listener.
-     */
-    String executorName();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/LocalNodeMasterListener.java:49]: Hurrah :tada:

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java (hunk 3 above)
Comment (jaymode) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java:135]: Great catch! Addressed in https://github.com/elastic/elasticsearch/pull/59932/commits/d30872145f375f109ed20e2d26db6946436c3b4e

**Comment by jaymode** [comment]
Thanks for the reviews @original-brownbear @dakrone @DaveCTurner

**Pull request merged**